Data Engineering Capstone Project
# Datasets Raw Exploration

In [1]:
!sh ../bootstrap_jupyter.sh

+ sudo apt-get update -y
Get:1 http://archive.ubuntu.com/ubuntu bionic InRelease [242 kB]
Get:2 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]      
Get:4 http://archive.ubuntu.com/ubuntu bionic-backports InRelease [74.6 kB]
Get:5 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]    
Get:6 http://archive.ubuntu.com/ubuntu bionic/main amd64 Packages [1,344 kB]   
Get:3 http://repos.mesosphere.com/ubuntu xenial InRelease [5,275 B]            
Get:7 http://archive.ubuntu.com/ubuntu bionic/multiverse amd64 Packages [186 kB]
Get:8 http://archive.ubuntu.com/ubuntu bionic/universe amd64 Packages [11.3 MB]
Get:9 http://archive.ubuntu.com/ubuntu bionic/restricted amd64 Packages [13.5 kB]
Get:10 http://archive.ubuntu.com/ubuntu bionic-updates/restricted amd64 Packages [73.6 kB]
Get:11 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 Packages [1,385 kB]
Get:12 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 Packages [1,226 kB]
Get:

Processing /home/anthelix/.cache/pip/wheels/f9/8d/8d/f6af3f7f9eea3553bc2fe6d53e4b287dad18b06a861ac56ddf/retrying-1.3.3-py3-none-any.whl
  Using cached s3transfer-0.3.3-py2.py3-none-any.whl (69 kB)
  Using cached botocore-1.16.16-py2.py3-none-any.whl (6.2 MB)
  Using cached jmespath-0.10.0-py2.py3-none-any.whl (24 kB)
  Using cached docutils-0.15.2-py3-none-any.whl (547 kB)
  Attempting uninstall: bokeh
    Found existing installation: bokeh 2.0.1
    Uninstalling bokeh-2.0.1:
      Successfully uninstalled bokeh-2.0.1
+ POSTGRES_JAR=postgresql-42.2.10.jar
+ [ -f postgresql-42.2.10.jar ]
+ echo postgresql-42.2.10.jar exist
postgresql-42.2.10.jar exist
+ sudo cp log4j.properties /usr/local/spark/conf/log4j.properties
cp: cannot stat 'log4j.properties': No such file or directory


In [2]:
import os
import sys
import datetime
import numpy as np
import pandas as pd
path = '../../data/'
pd.set_option('display.max_columns', None)

In [3]:
list_filename= ['I94_SAS_Labels_Descriptions.SAS', 'immigration_data_sample.csv', 
                "GlobalLandTemperaturesByCity.csv", 'wikipedia-iso-country-codes.csv',
                "airports-extended.csv","airport-codes_csv.csv", 'us-cities-demographics.csv',
               'WDIData.csv', "education-statistics/EdStatsData.csv"]

In [4]:
def sum1forline(list_filename):
    my_sum = 0
    path = '../../data/'
    for filename in list_filename:
        with open(path+filename) as f:
            a = sum(1 for line in f)
            print("{} rows in {} .".format(a, filename))
            my_sum+=a
            
    return("Total of the files : {:,}".format(my_sum))

In [5]:
sum1forline(list_filename)

1100 rows in I94_SAS_Labels_Descriptions.SAS .
1001 rows in immigration_data_sample.csv .
8599213 rows in GlobalLandTemperaturesByCity.csv .
247 rows in wikipedia-iso-country-codes.csv .
10668 rows in airports-extended.csv .
55076 rows in airport-codes_csv.csv .
2892 rows in us-cities-demographics.csv .
422137 rows in WDIData.csv .
886931 rows in education-statistics/EdStatsData.csv .


'Total of the files : 9,979,265'

## I94 Immigration Data

* immigration_data_sample.csv
* There are 1000 rows and 29 columns in *immigration_data_sample.csv*.
* df_immigration

In [6]:
file = "immigration_data_sample.csv"
df = pd.read_csv(path+file, sep=",")
nRow, nCol = df.shape
print("There are {} rows and {} columns in  *{}*. ".format(nRow, nCol, file))
display(df.dropna(subset=["insnum"]).head(2))
tab_info=pd.DataFrame(df.dtypes).T.rename(index={0:'column type'})
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()).T.rename(index={0:'null values (nb)'}))
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()/df.shape[0]*100)
                         .T.rename(index={0:'null values (%)'}))
display(tab_info)
#change subset

#display(df[pd.notnull(df).any(axis=1)].head(2))

There are 1000 rows and 29 columns in  *immigration_data_sample.csv*. 


Unnamed: 0      cicid   i94yr  i94mon  i94cit  i94res i94port  arrdate  \
15     2938436  5960799.0  2016.0     4.0   245.0   245.0     SAI  20545.0   
35     3020206  6055103.0  2016.0     4.0   252.0   209.0     AGA  20572.0   

    i94mode i94addr  depdate  i94bir  i94visa  count  dtadfile visapost occup  \
15      1.0     NaN  20550.0    30.0      2.0    1.0  20160615      NaN   NaN   
35      1.0      GU  20577.0    27.0      2.0    1.0  20160801      NaN   NaN   

   entdepa entdepd  entdepu matflag  biryear   dtaddto gender  insnum airline  \
15       P       D      NaN       M   1986.0  04132016      M  3882.0      MU   
35       A       D      NaN       M   1989.0  06112016      M  3943.0      TW   

          admnum  fltno visatype  
15  4.416258e+10  00763       CP  
35  5.109649e+10  00311      GMT

Unnamed: 0    cicid    i94yr   i94mon   i94cit   i94res  \
column type           int64  float64  float64  float64  float64  float64   
null values (nb)          0        0        0        0        0        0   
null values (%)           0        0        0        0        0        0   

                 i94port  arrdate  i94mode i94addr  depdate   i94bir  i94visa  \
column type       object  float64  float64  object  float64  float64  float64   
null values (nb)       0        0        0      59       49        0        0   
null values (%)        0        0        0     5.9      4.9        0        0   

                    count dtadfile visapost   occup entdepa entdepd  entdepu  \
column type       float64    int64   object  object  object  object  float64   
null values (nb)        0        0      618     996       0      46     1000   
null values (%)         0        0     61.8    99.6       0     4.6      100   

                 matflag  biryear dtaddto  gender   insnum airline   admnum  \
column type       object  float64  object  object  float64  object  float64   
null values (nb)      46        0       0     141      965      33        0   
null values (%)      4.6        0       0    14.1     96.5     3.3        0   

                   fltno visatype  
column type       object   object  
null values (nb)       8        0  
null values (%)      0.8        0

## Global Land Temperature Data
* GlobalLandTemperaturesByCity.csv
* There are 8599212 rows and 7 columns in  *GlobalLandTemperaturesByCity.csv*. 
* df_temperature

In [7]:
file = "GlobalLandTemperaturesByCity.csv"
df = pd.read_csv(path+file, sep=",", nrows=9000000)
nRow, nCol = df.shape
print("There are {} rows and {} columns in  *{}*. ".format(nRow, nCol, file))

display(df.head(2))
tab_info=pd.DataFrame(df.dtypes).T.rename(index={0:'column type'})
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()).T.rename(index={0:'null values (nb)'}))
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()/df.shape[0]*100)
                         .T.rename(index={0:'null values (%)'}))
tab_info
#change subset
#df.dropna(subset=["insnum"]).head(2)

There are 8599212 rows and 7 columns in  *GlobalLandTemperaturesByCity.csv*. 


dt  AverageTemperature  AverageTemperatureUncertainty   City  \
0  1743-11-01               6.068                          1.737  Århus   
1  1743-12-01                 NaN                            NaN  Århus   

   Country Latitude Longitude  
0  Denmark   57.05N    10.33E  
1  Denmark   57.05N    10.33E

dt AverageTemperature AverageTemperatureUncertainty  \
column type       object            float64                       float64   
null values (nb)       0             364130                        364130   
null values (%)        0            4.23446                       4.23446   

                    City Country Latitude Longitude  
column type       object  object   object    object  
null values (nb)       0       0        0         0  
null values (%)        0       0        0         0

## Airports Code Data
* airport-codes_csv.csv
* There are 55075 rows and 12 columns in  *airport-codes_csv.csv*.
* df_airport_code

In [8]:
file = "airport-codes_csv.csv"
df = pd.read_csv(path+file, sep=",", nrows=9000000)
nRow, nCol = df.shape
print("There are {} rows and {} columns in  *{}*. ".format(nRow, nCol, file))
#display(df[pd.notnull(df).any(axis=1)].head(2))
#display(df.head(2))
display(df.dropna(subset=["iata_code"]).head(2))
tab_info=pd.DataFrame(df.dtypes).T.rename(index={0:'column type'})
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()).T.rename(index={0:'null values (nb)'}))
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()/df.shape[0]*100)
                         .T.rename(index={0:'null values (%)'}))
tab_info
#change subse

There are 55075 rows and 12 columns in  *airport-codes_csv.csv*. 


ident           type                     name  elevation_ft continent  \
223   03N  small_airport           Utirik Airport           4.0        OC   
440  07FA  small_airport  Ocean Reef Club Airport           8.0       NaN   

    iso_country iso_region   municipality gps_code iata_code local_code  \
223          MH     MH-UTI  Utirik Island     K03N       UTK        03N   
440          US      US-FL      Key Largo     07FA       OCA       07FA   

                           coordinates  
223                 169.852005, 11.222  
440  -80.274803161621, 25.325399398804

ident    type    name elevation_ft continent iso_country  \
column type       object  object  object      float64    object      object   
null values (nb)       0       0       0         7006     27719         247   
null values (%)        0       0       0      12.7208   50.3296    0.448479   

                 iso_region municipality gps_code iata_code local_code  \
column type          object       object   object    object     object   
null values (nb)          0         5676    14045     45886      26389   
null values (%)           0      10.3059  25.5016   83.3155    47.9147   

                 coordinates  
column type           object  
null values (nb)           0  
null values (%)            0

## Global Airports Data
* airports-extended.csv
* There are 10668 rows and 13 columns in  *airports-extended.csv*. 
* df_global_airports

In [9]:
file = "airports-extended.csv"
df = pd.read_csv(path+file, 
                 header = None, 
                 index_col=0,
                 names=['id', 'name', 'city', 'country', 'iata', 'icao', 'latitude', 'longitude', 'altitude', 
                            'timezone', 'dst', 'tz_timezone', 'type', 'data_source']).rename(columns=lambda col: col)

nRow, nCol = df.shape
print("There are {} rows and {} columns in  *{}*. ".format(nRow, nCol, file))
#display(df[pd.notnull(df).any(axis=1)].head(2))
display(df.head(2))
tab_info=pd.DataFrame(df.dtypes).T.rename(index={0:'column type'})
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()).T.rename(index={0:'null values (nb)'}))
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()/df.shape[0]*100)
                         .T.rename(index={0:'null values (%)'}))
tab_info
#change subset
#df.dropna(subset=["insnum"]).head(2)

There are 10668 rows and 13 columns in  *airports-extended.csv*. 


name    city           country iata  icao  latitude   longitude  \
id                                                                              
1   Goroka Airport  Goroka  Papua New Guinea  GKA  AYGA  -6.08169  145.391998   
2   Madang Airport  Madang  Papua New Guinea  MAG  AYMD  -5.20708  145.789001   

    altitude timezone dst           tz_timezone     type  data_source  
id                                                                     
1       5282       10   U  Pacific/Port_Moresby  airport  OurAirports  
2         20       10   U  Pacific/Port_Moresby  airport  OurAirports

name     city country       iata        icao latitude  \
column type       object   object  object     object      object  float64   
null values (nb)       0       47       0          2           1        0   
null values (%)        0  0.44057       0  0.0187477  0.00937383        0   

                 longitude altitude timezone     dst tz_timezone    type  \
column type        float64    int64   object  object      object  object   
null values (nb)         0        0        0       0           0       0   
null values (%)          0        0        0       0           0       0   

                 data_source  
column type           object  
null values (nb)           0  
null values (%)            0

## Iso Country Data
* wikipedia-iso-country-codes.csv
* There are 246 rows and 4 columns in  *wikipedia-iso-country-codes.csv*.
* df_iso_country

In [10]:
file = "wikipedia-iso-country-codes.csv"
df = pd.read_csv(path+file, sep=",", nrows=9000000)
nRow, nCol = df.shape
print("There are {} rows and {} columns in  *{}*. ".format(nRow, nCol, file))
#display(df[pd.notnull(df).any(axis=1)].head(2))
display(df.head(2))
tab_info=pd.DataFrame(df.dtypes).T.rename(index={0:'column type'})
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()).T.rename(index={0:'null values (nb)'}))
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()/df.shape[0]*100)
                         .T.rename(index={0:'null values (%)'}))
tab_info
#change subset
#df.dropna(subset=["insnum"]).head(2)

There are 246 rows and 5 columns in  *wikipedia-iso-country-codes.csv*. 


English short name lower case Alpha-2 code Alpha-3 code  Numeric code  \
0                      Zimbabwe           ZW          ZWE           716   
1                        Zambia           ZM          ZMB           894   

      ISO 3166-2  
0  ISO 3166-2:ZW  
1  ISO 3166-2:ZM

English short name lower case Alpha-2 code Alpha-3 code  \
column type                             object       object       object   
null values (nb)                             0            1            0   
null values (%)                              0     0.406504            0   

                 Numeric code ISO 3166-2  
column type             int64     object  
null values (nb)            0          0  
null values (%)             0          0

## US cities Demographics
* us-cities-demographics.csv
* There are 2891 rows and 12 columns in  *us-cities-demographics.csv*. 
* df_demograph

In [11]:
file = "us-cities-demographics.csv"
df = pd.read_csv(path+file, sep=";")
nRow, nCol = df.shape
print("There are {} rows and {} columns in  *{}*. ".format(nRow, nCol, file))

#display(df[pd.notnull(df).any(axis=1)].head(2))
display(df.head(2))
#display(df.dropna(subset=["iata_code"]).head(2))

tab_info=pd.DataFrame(df.dtypes).T.rename(index={0:'column type'})
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()).T.rename(index={0:'null values (nb)'}))
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()/df.shape[0]*100)
                         .T.rename(index={0:'null values (%)'}))
tab_info
#change subset

There are 2891 rows and 12 columns in  *us-cities-demographics.csv*. 


City          State  Median Age  Male Population  \
0  Silver Spring       Maryland        33.8          40601.0   
1         Quincy  Massachusetts        41.0          44129.0   

   Female Population  Total Population  Number of Veterans  Foreign-born  \
0            41862.0             82463              1562.0       30908.0   
1            49500.0             93629              4147.0       32935.0   

   Average Household Size State Code                Race  Count  
0                    2.60         MD  Hispanic or Latino  25924  
1                    2.39         MA               White  58723

City   State Median Age Male Population Female Population  \
column type       object  object    float64         float64           float64   
null values (nb)       0       0          0               3                 3   
null values (%)        0       0          0         0.10377           0.10377   

                 Total Population Number of Veterans Foreign-born  \
column type                 int64            float64      float64   
null values (nb)                0                 13           13   
null values (%)                 0           0.449671     0.449671   

                 Average Household Size State Code    Race  Count  
column type                     float64     object  object  int64  
null values (nb)                     16          0       0      0  
null values (%)                0.553442          0       0      0

## World Development Indicators Data
* WDIData.csv
* There are 422136 rows and 64 columns in  *WDIData.csv*.
* df_indicator_dev

In [12]:
file = "WDIData.csv"
df = pd.read_csv(path+file)
nRow, nCol = df.shape
print("There are {} rows and {} columns in  *{}*. ".format(nRow, nCol, file))

#display(df[pd.notnull(df).any(axis=1)].head(2))
display(df.head(2))
#display(df.dropna(subset=["iata_code"]).head(2))

tab_info=pd.DataFrame(df.dtypes).T.rename(index={0:'column type'})
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()).T.rename(index={0:'null values (nb)'}))
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()/df.shape[0]*100)
                         .T.rename(index={0:'null values (%)'}))
tab_info
#change subset

There are 422136 rows and 64 columns in  *WDIData.csv*. 


Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   

                                      Indicator Name     Indicator Code  1960  \
0  2005 PPP conversion factor, GDP (LCU per inter...      PA.NUS.PPP.05   NaN   
1  2005 PPP conversion factor, private consumptio...  PA.NUS.PRVT.PP.05   NaN   

   1961  1962  1963  1964  1965  1966  1967  1968  1969  1970  1971  1972  \
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

   1973  1974  1975  1976  1977  1978  1979  1980  1981  1982  1983  1984  \
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

   1985  1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  1996  \
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

   1997  1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  2008  \
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

   2009  2010  2011  2012  2013  2014  2015  2016  2017  2018  Unnamed: 63  
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN

Country Name Country Code Indicator Name Indicator Code  \
column type            object       object         object         object   
null values (nb)            0            0              0              0   
null values (%)             0            0              0              0   

                     1960     1961     1962     1963     1964     1965  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   383874   380159   377980   378100   377572   374886   
null values (%)   90.9361   90.056  89.5399  89.5683  89.4432  88.8069   

                     1966     1967     1968     1969     1970     1971  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   375289   373557   373978   372175   329188   322941   
null values (%)   88.9024  88.4921  88.5918  88.1647  77.9815  76.5016   

                     1972     1973     1974     1975     1976     1977  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   320189   320579   319205   315106   313014   309504   
null values (%)   75.8497  75.9421  75.6166  74.6456    74.15  73.3186   

                     1978     1979     1980     1981     1982     1983  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   309893   309060   304245   302355   301070   301293   
null values (%)   73.4107  73.2134  72.0727   71.625  71.3206  71.3734   

                     1984     1985     1986     1987     1988     1989  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   300466   299300   298440   297044   297320   295547   
null values (%)   71.1775  70.9013  70.6976  70.3669  70.4323  70.0123   

                     1990     1991     1992     1993     1994     1995  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   271268   266167   262167   260677   258472   250812   
null values (%)   64.2608  63.0524  62.1049  61.7519  61.2296   59.415   

                     1996     1997     1998     1999     2000     2001  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   250584   249441   247381   242129   217939   227649   
null values (%)    59.361  59.0902  58.6022  57.3581  51.6277  53.9279   

                     2002     2003     2004     2005     2006     2007  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   222698   221900   216441   202656   204650   200703   
null values (%)    52.755   52.566  51.2728  48.0073  48.4796  47.5446   

                     2008     2009     2010     2011     2012     2013  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   201672   200654   185174   194482   192860   197314   
null values (%)   47.7742   47.533   43.866  46.0709  45.6867  46.7418   

                     2014     2015     2016     2017     2018 Unnamed: 63  
column type       float64  float64  float64  float64  float64     float64  
null values (nb)   192098   198058   206444   232401   338818      422136  
null values (%)   45.5062  46.9181  48.9046  55.0536  80.2628         100

## Education statistics Data
* education-statistics/EdStatsData.csv
* There are 886930 rows and 70 columns in  *education-statistics/EdStatsData.csv*. 
* df_Educ_data
* Edit: i find some data in 'World Development Indicators Data' so I didn't use this dataset

In [13]:
file = "education-statistics/EdStatsData.csv"
df = pd.read_csv(path+file)
nRow, nCol = df.shape
print("There are {} rows and {} columns in  *{}*. ".format(nRow, nCol, file))

#display(df[pd.notnull(df).any(axis=1)].head(2))
display(df.head(2))
#display(df.dropna(subset=["iata_code"]).head(2))

tab_info=pd.DataFrame(df.dtypes).T.rename(index={0:'column type'})
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()).T.rename(index={0:'null values (nb)'}))
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()/df.shape[0]*100)
                         .T.rename(index={0:'null values (%)'}))
tab_info

There are 886930 rows and 70 columns in  *education-statistics/EdStatsData.csv*. 


Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   

                                      Indicator Name Indicator Code  1970  \
0  Adjusted net enrolment rate, lower secondary, ...     UIS.NERA.2   NaN   
1  Adjusted net enrolment rate, lower secondary, ...   UIS.NERA.2.F   NaN   

   1971  1972  1973  1974  1975  1976  1977  1978  1979  1980  1981  1982  \
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

   1983  1984  1985  1986  1987  1988  1989  1990  1991  1992  1993  1994  \
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

   1995  1996  1997  1998  1999  2000  2001  2002  2003  2004  2005  2006  \
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

   2007  2008  2009  2010  2011  2012  2013  2014  2015  2016  2017  2020  \
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

   2025  2030  2035  2040  2045  2050  2055  2060  2065  2070  2075  2080  \
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

   2085  2090  2095  2100  Unnamed: 69  
0   NaN   NaN   NaN   NaN          NaN  
1   NaN   NaN   NaN   NaN          NaN

Country Name Country Code Indicator Name Indicator Code  \
column type            object       object         object         object   
null values (nb)            0            0              0              0   
null values (%)             0            0              0              0   

                     1970     1971     1972     1973     1974     1975  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   814642   851393   851311   851385   851200   799624   
null values (%)   91.8496  95.9933   95.984  95.9924  95.9715  90.1564   

                     1976     1977     1978     1979     1980     1981  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   849447   849356   849354   850121   797808   848153   
null values (%)   95.7738  95.7636  95.7634  95.8498  89.9516   95.628   

                     1982     1983     1984     1985     1986     1987  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   849419   848470   848324   796634   847558   848289   
null values (%)   95.7707  95.6637  95.6472  89.8193  95.5609  95.6433   

                     1988     1989     1990     1991     1992     1993  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   848378   849390   762525   812493   811387   811137   
null values (%)   95.6533  95.7674  85.9735  91.6073  91.4826  91.4545   

                     1994     1995     1996     1997     1998     1999  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   809468   755569   810123   813477   802016   768091   
null values (%)   91.2663  85.1892  91.3401  91.7183  90.4261  86.6011   

                     2000     2001     2002     2003     2004     2005  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   710254   763421   762725   756567   758116   702822   
null values (%)   80.0801  86.0745  85.9961  85.3018  85.4764  79.2421   

                     2006     2007     2008     2009     2010     2011  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   746618   749658   752543   744822   644488   740918   
null values (%)     84.18  84.5228  84.8481  83.9775   72.665  83.5374   

                     2012     2013     2014     2015     2016     2017  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   739666   749421   773141   755872   870470   886787   
null values (%)   83.3962  84.4961  87.1705  85.2234  98.1442  99.9839   

                     2020     2025     2030     2035     2040     2045  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   835494   835494   835494   835494   835494   835494   
null values (%)   94.2007  94.2007  94.2007  94.2007  94.2007  94.2007   

                     2050     2055     2060     2065     2070     2075  \
column type       float64  float64  float64  float64  float64  float64   
null values (nb)   835494   835494   835494   835494   835494   835494   
null values (%)   94.2007  94.2007  94.2007  94.2007  94.2007  94.2007   

                     2080     2085     2090     2095     2100 Unnamed: 69  
column type       float64  float64  float64  float64  float64     float64  
null values (nb)   835494   835494   835494   835494   835494      886930  
null values (%)   94.2007  94.2007  94.2007  94.2007  94.2007         100

In [14]:
file = "i94addr.csv"
df = pd.read_csv(path+file)
nRow, nCol = df.shape
print("There are {} rows and {} columns in  *{}*. ".format(nRow, nCol, file))

#display(df[pd.notnull(df).any(axis=1)].head(2))
display(df.head(2))
#display(df.dropna(subset=["iata_code"]).head(2))

tab_info=pd.DataFrame(df.dtypes).T.rename(index={0:'column type'})
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()).T.rename(index={0:'null values (nb)'}))
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()/df.shape[0]*100)
                         .T.rename(index={0:'null values (%)'}))
tab_info
#change subset

There are 55 rows and 2 columns in  *i94addr.csv*. 


State_id    State
0       Al  Alabama
1       Ak   Alaska

State_id   State
column type        object  object
null values (nb)        0       0
null values (%)         0       0

In [15]:
file = "i94mode.csv"
df = pd.read_csv(path+file)
nRow, nCol = df.shape
print("There are {} rows and {} columns in  *{}*. ".format(nRow, nCol, file))

#display(df[pd.notnull(df).any(axis=1)].head(2))
display(df.head(2))
#display(df.dropna(subset=["iata_code"]).head(2))

tab_info=pd.DataFrame(df.dtypes).T.rename(index={0:'column type'})
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()).T.rename(index={0:'null values (nb)'}))
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()/df.shape[0]*100)
                         .T.rename(index={0:'null values (%)'}))
tab_info
#change subset

There are 4 rows and 2 columns in  *i94mode.csv*. 


Mode_id Mode
0        1  Air
1        2  Sea

Mode_id    Mode
column type        int64  object
null values (nb)       0       0
null values (%)        0       0

In [16]:
file = "i94port.csv"
df = pd.read_csv(path+file)
nRow, nCol = df.shape
print("There are {} rows and {} columns in  *{}*. ".format(nRow, nCol, file))

#display(df[pd.notnull(df).any(axis=1)].head(2))
display(df.head(2))
#display(df.dropna(subset=["iata_code"]).head(2))

tab_info=pd.DataFrame(df.dtypes).T.rename(index={0:'column type'})
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()).T.rename(index={0:'null values (nb)'}))
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()/df.shape[0]*100)
                         .T.rename(index={0:'null values (%)'}))
tab_info
#change subset

There are 583 rows and 3 columns in  *i94port.csv*. 


Port_id  Port_city State_id
0     Alc      Alcan       Ak
1     Anc  Anchorage       Ak

Port_id Port_city State_id
column type       object    object   object
null values (nb)       0         0        0
null values (%)        0         0        0

In [17]:
file = "i94visa.csv"
df = pd.read_csv(path+file)
nRow, nCol = df.shape
print("There are {} rows and {} columns in  *{}*. ".format(nRow, nCol, file))

#display(df[pd.notnull(df).any(axis=1)].head(2))
display(df.head(2))
#display(df.dropna(subset=["iata_code"]).head(2))

tab_info=pd.DataFrame(df.dtypes).T.rename(index={0:'column type'})
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()).T.rename(index={0:'null values (nb)'}))
tab_info=tab_info.append(pd.DataFrame(df.isnull().sum()/df.shape[0]*100)
                         .T.rename(index={0:'null values (%)'}))
tab_info
#change subset

There are 3 rows and 2 columns in  *i94visa.csv*. 


Code_visa      Visa
0          1  Business
1          2  Pleasure

Code_visa    Visa
column type          int64  object
null values (nb)         0       0
null values (%)          0       0